cthread/src/lib.c

```diff
@@ -475,7 +476,120 @@ int cwait(csem_t *sem)
 
 int csignal(csem_t *sem)
 {
-	return -1;
+	//incrementa o contador para indicar que o recurso foi liberado
+	sem->count = sem->count + 1;
+	
+	//procura uma thread de alta prioridade para desalocar da fila do semáforo
+	int found =0;
+	TCB_t *ThreadFromSem;
+	FirstFila2(sem->fila); 
+	
+	do
+	{
+		ThreadFromSem = (TCB_t *)GetAtIteratorFila2(sem->fila);
+		if (ThreadFromSem != NULL && ThreadFromSem->prio == ALTA_PRIORIDADE)
+		{
+			found = 1;
+		}
+	} while (!NextFila2(sem->fila) && !found);
+	if (!found)
+	{
+		//se não houver uma de alta prioridade procura uma thread de media prioridade na fila do semáforo
+		FirstFila2(sem->fila); 
+		do
+		{
+			ThreadFromSem = (TCB_t *)GetAtIteratorFila2(sem->fila);
+			if (ThreadFromSem != NULL && ThreadFromSem->prio == ALTA_PRIORIDADE)
+			{
+				found = 1;
+			}
+		} while (!NextFila2(sem->fila) && !found);
+		if(!found)
+		{
+			// se não houver de media prioridade procura uma de baixa prioridade na fila do semáforo
+			FirstFila2(sem->fila); 
+			do
+			{
+				ThreadFromSem = (TCB_t *)GetAtIteratorFila2(sem->fila);
+				if(ThreadFromSem != NULL)
+				{
+					found =1;
+				}
+			} while(!NextFila2(sem->fila) && !found);
+			if(!found)
+				printf(" Nao tem nenhuma thread na fila do semaforo \n");
+		
+		}
+	}
+	// coloca a thread em alguma fila de aptos
+	
+	if(found)
+	{	
+		ThreadFromSem->state = PROCST_APTO;
+		setIteratorToFirst();
+		
+		switch (ThreadFromSem->prio)
+		{
+			case BAIXA_PRIORIDADE:
+				if (AppendFila2(pfilaAptoBaixa, (void *)ThreadFromSem) != 0)
+				{
+					printf(" func csignal: Erro ao inserir a thread na fila de aptos");
+					return -1;
```

Comment:
please change all hard coded values to defined constants, this way it is easier to debug